File: src/services/validation.js
// src/services/validation.js
// Schémas de validation Yup pour tous les formulaires

import * as Yup from 'yup';

/**
 * Validation Login
 */
export const LoginSchema = Yup.object().shape({
  username: Yup.string()
    .min(3, 'Minimum 3 caractères')
    .max(50, 'Maximum 50 caractères')
    .required('Nom d\'utilisateur requis'),
  password: Yup.string()
    .min(6, 'Minimum 6 caractères')
    .required('Mot de passe requis'),
});

/**
 * Validation Client
 */
export const ClientSchema = Yup.object().shape({
  name: Yup.string()
    .min(2, 'Minimum 2 caractères')
    .max(100, 'Maximum 100 caractères')
    .required('Nom requis'),
  email: Yup.string()
    .email('Email invalide')
    .required('Email requis'),
  phone: Yup.string()
    .matches(/^[\d\s\-\+\(\)]{10,}$/, 'Téléphone invalide')
    .required('Téléphone requis'),
  address: Yup.string()
    .max(255, 'Maximum 255 caractères')
    .required('Adresse requise'),
});

/**
 * Validation Produit
 */
export const ProductSchema = Yup.object().shape({
  name: Yup.string()
    .min(2, 'Minimum 2 caractères')
    .max(100, 'Maximum 100 caractères')
    .required('Nom requis'),
  barcode: Yup.string()
    .max(50, 'Code barre invalide')
    .required('Code barre requis'),
  category: Yup.string()
    .required('Catégorie requise'),
  price: Yup.number()
    .positive('Le prix doit être positif')
    .required('Prix requis'),
  stock_quantity: Yup.number()
    .min(0, 'Stock ne peut être négatif')
    .required('Stock requis'),
  min_stock: Yup.number()
    .min(0, 'Stock minimum ne peut être négatif'),
  description: Yup.string()
    .max(500, 'Maximum 500 caractères'),
});

/**
 * Validation Vente/Ligne de vente
 */
export const SaleLineSchema = Yup.object().shape({
  product_id: Yup.number()
    .positive()
    .required('Produit requis'),
  quantity: Yup.number()
    .positive('Quantité doit être positive')
    .required('Quantité requise'),
  unit_price: Yup.number()
    .positive('Prix doit être positif')
    .required('Prix requis'),
});

export const SaleSchema = Yup.object().shape({
  client_id: Yup.number()
    .positive()
    .required('Client requis'),
  lines: Yup.array()
    .of(SaleLineSchema)
    .min(1, 'Minimum 1 ligne requis'),
  notes: Yup.string()
    .max(500, 'Maximum 500 caractères'),
});

/**
 * Validation Config API
 */
export const APIConfigSchema = Yup.object().shape({
  API_URL: Yup.string()
    .url('URL invalide')
    .required('URL API requise'),
  API_TIMEOUT: Yup.number()
    .positive()
    .required('Timeout requis'),
});

/**
 * Validation Utilisateur (création)
 */
export const UserSchema = Yup.object().shape({
  username: Yup.string()
    .min(3, 'Minimum 3 caractères')
    .max(50, 'Maximum 50 caractères')
    .matches(/^[a-zA-Z0-9_-]+$/, 'Caractères autorisés: lettres, chiffres, tiret, underscore')
    .required('Nom d\'utilisateur requis'),
  password: Yup.string()
    .min(6, 'Minimum 6 caractères')
    .max(100, 'Maximum 100 caractères')
    .required('Mot de passe requis'),
  fullname: Yup.string()
    .max(100, 'Maximum 100 caractères'),
  role: Yup.string()
    .oneOf(['user', 'admin'], 'Rôle invalide')
    .required('Rôle requis'),
});

/**
 * Validation Changement de mot de passe
 */
export const UserPasswordSchema = Yup.object().shape({
  password: Yup.string()
    .min(6, 'Minimum 6 caractères')
    .max(100, 'Maximum 100 caractères')
    .required('Mot de passe requis'),
});

/**
 * Utilitaire: valider et retourner les erreurs
 */
export const validateForm = async (schema, values) => {
  try {
    await schema.validate(values, { abortEarly: false });
    return { isValid: true, errors: {} };
  } catch (err) {
    const errors = {};
    if (err.inner) {
      err.inner.forEach((fieldError) => {
        errors[fieldError.path] = fieldError.message;
      });
    }
    return { isValid: false, errors };
  }
};

File: backend/validation.py
# backend/validation.py
# Validation des inputs

import re
from typing import Dict, List, Tuple


def validate_email(email: str) -> Tuple[bool, str]:
    """Valider une adresse email"""
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if not email or not re.match(pattern, email):
        return False, 'Email invalide'
    return True, ''


def validate_phone(phone: str) -> Tuple[bool, str]:
    """Valider un numéro de téléphone"""
    # Simple: au moins 10 chiffres
    digits = re.sub(r'\D', '', phone)
    if len(digits) < 10:
        return False, 'Téléphone invalide (min 10 chiffres)'
    return True, ''


def validate_username(username: str) -> Tuple[bool, str]:
    """Valider un nom d'utilisateur"""
    if len(username) < 3:
        return False, 'Minimum 3 caractères'
    if len(username) > 50:
        return False, 'Maximum 50 caractères'
    if not re.match(r'^[a-zA-Z0-9_-]+$', username):
        return False, 'Caractères valides: a-z, A-Z, 0-9, _, -'
    return True, ''


def validate_password(password: str) -> Tuple[bool, str]:
    """Valider un mot de passe"""
    if len(password) < 6:
        return False, 'Minimum 6 caractères'
    if len(password) > 100:
        return False, 'Maximum 100 caractères'
    return True, ''


def validate_product_data(data: Dict) -> Tuple[bool, Dict]:
    """Valider les données d'un produit"""
    errors = {}
    
    # Nom
    name = data.get('name', '').strip()
    if not name or len(name) < 2 or len(name) > 100:
        errors['name'] = 'Nom: 2-100 caractères'
    
    # Code barre
    barcode = data.get('barcode', '').strip()
    if not barcode or len(barcode) > 50:
        errors['barcode'] = 'Code barre invalide'
    
    # Prix
    try:
        price = float(data.get('price', 0))
        if price <= 0:
            errors['price'] = 'Prix doit être positif'
    except (ValueError, TypeError):
        errors['price'] = 'Prix invalide'
    
    # Stock
    try:
        stock = int(data.get('stock_quantity', 0))
        if stock < 0:
            errors['stock_quantity'] = 'Stock ne peut être négatif'
    except (ValueError, TypeError):
        errors['stock_quantity'] = 'Stock invalide'
    
    return len(errors) == 0, errors


def validate_client_data(data: Dict) -> Tuple[bool, Dict]:
    """Valider les données d'un client"""
    errors = {}
    
    # Nom
    name = data.get('name', '').strip()
    if not name or len(name) < 2 or len(name) > 100:
        errors['name'] = 'Nom: 2-100 caractères'
    
    # Email
    email = data.get('email', '').strip()
    if email:
        is_valid, msg = validate_email(email)
        if not is_valid:
            errors['email'] = msg
    
    # Téléphone
    phone = data.get('phone', '').strip()
    if phone:
        is_valid, msg = validate_phone(phone)
        if not is_valid:
            errors['phone'] = msg
    
    return len(errors) == 0, errors


def sanitize_string(s: str, max_length: int = 255) -> str:
    """Nettoyer une chaîne"""
    if not isinstance(s, str):
        return ''
    return s.strip()[:max_length]
